beacon-chain/rpc/eth/light-client/handlers.go

```diff
@@ -292,72 +283,58 @@ func (s *Server) GetLightClientFinalityUpdate(w http.ResponseWriter, req *http.R
 	if !shared.WriteBlockFetchError(w, block, err) {
 		return
 	}
-
-	state, err := s.Stater.StateBySlot(ctx, block.Block().Slot())
+	st, err := s.Stater.StateBySlot(ctx, block.Block().Slot())
 	if err != nil {
 		httputil.HandleError(w, "could not get state: "+err.Error(), http.StatusInternalServerError)
 		return
 	}
-
-	// Get attested state
 	attestedRoot := block.Block().ParentRoot()
 	attestedBlock, err := s.Blocker.Block(ctx, attestedRoot[:])
 	if err != nil {
-		httputil.HandleError(w, "could not get attested block: "+err.Error(), http.StatusInternalServerError)
+		httputil.HandleError(w, "Could not get attested block: "+err.Error(), http.StatusInternalServerError)
 		return
 	}
 	if attestedBlock == nil {
-		httputil.HandleError(w, "attested block is nil", http.StatusInternalServerError)
+		httputil.HandleError(w, "Attested block is nil", http.StatusInternalServerError)
 		return
 	}
-
 	attestedSlot := attestedBlock.Block().Slot()
 	attestedState, err := s.Stater.StateBySlot(ctx, attestedSlot)
 	if err != nil {
-		httputil.HandleError(w, "could not get attested state: "+err.Error(), http.StatusInternalServerError)
+		httputil.HandleError(w, "Could not get attested state: "+err.Error(), http.StatusInternalServerError)
 		return
 	}
 
-	update, err := newLightClientOptimisticUpdateFromBeaconState(
-		ctx,
-		state,
-		block,
-		attestedState,
-	)
+	update, err := newLightClientOptimisticUpdateFromBeaconState(ctx, st, block, attestedState)
 	if err != nil {
-		httputil.HandleError(w, "could not get light client optimistic update: "+err.Error(), http.StatusInternalServerError)
+		httputil.HandleError(w, "Could not get light client optimistic update: "+err.Error(), http.StatusInternalServerError)
 		return
 	}
 
-	response := &structs.LightClientUpdateWithVersion{
+	response := &structs.LightClientOptimisticUpdateResponse{
 		Version: version.String(attestedState.Version()),
 		Data:    update,
 	}
 
 	httputil.WriteJson(w, response)
 }
 
-// getLightClientEventBlock - returns the block that should be used for light client events, which satisfies the minimum number of signatures from sync committee
-func (s *Server) getLightClientEventBlock(ctx context.Context, minSignaturesRequired uint64) (interfaces.ReadOnlySignedBeaconBlock, error) {
-	// Get the current state
-	state, err := s.HeadFetcher.HeadState(ctx)
+// suitableBlock returns the latest block that satisfies all criteria required for creating a new update
+func (s *Server) suitableBlock(ctx context.Context, minSignaturesRequired uint64) (interfaces.ReadOnlySignedBeaconBlock, error) {
+	st, err := s.HeadFetcher.HeadState(ctx)
 	if err != nil {
 		return nil, errors.Wrap(err, "could not get head state")
 	}
 
-	// Get the block
-	latestBlockHeader := *state.LatestBlockHeader()
-	stateRoot, err := state.HashTreeRoot(ctx)
+	latestBlockHeader := st.LatestBlockHeader()
+	stateRoot, err := st.HashTreeRoot(ctx)
```

Comment:
I wonder how long this will take

Reply:
I believe it shouldn't be too bad since it's the head state (meaning that it's already in memory and probably has most, if not all, field roots cached) and hashing the state is very optimized.